solutions/CF Packmen.mdx

```diff
@@ -0,0 +1,295 @@
+---
+id: cf-packmen
+source: CF
+title: Packmen
+author: Kevin Sheng
+---
+
+(No official editorial)
+
+Binary search on the minimum time it will take for all the Packmen to eat all the pellets.
+For each Packman, we can greedily calculate the maximum number of pellets that it can eat in the given amount of time and then remove those pellets.
+However, each Packman must eat all of the uneaten food to its left, or else all the food can't be eaten in the given time. Assigning other Packmen to the uneaten pellets would be pointless, because they would be farther away and thus have no way of reaching the uneatable pellet.
```

Comment:
Still quite unclear in my opinion.

First sentence can be "For each Packman, we can greedily assign the pellets it will eat"

The second part needs to be proof-like (to a certain extent) and articulate. I know this is difficult to do, but I think it's really beneficial. The way that FHC did it isn't optimal either, but it much more rigorous.

Key points for one possible explanation (similar to FHC):
 * Time only depends on leftmost and rightmost pellets. Therefore, every packman will eat a contiguous, disjoint subsegment
 * Prove that it is never less beneficial to assign packmen to subsegments in sorted order (with care, there are examples of subsegments that require packmen to be assigned in un-sorted order, but these can always be modified to work with sorted order)

Reply:
What's FHC?

Reply:
facebook hacker cup. Here's [link](https://www.facebook.com/codingcompetitions/hacker-cup/2020/round-1/problems/B/solution)

Reply:
>  there are examples of subsegments that require packmen to be assigned in un-sorted order

Could I have a small example?

Reply:
Packmen at positions 2 and 4.
Food at positions 0, 2, 6.

Assign packman 2 -> foods at 2 and 6
Assign packman 4 -> food at 0
The minimum # of steps is 4

You cannot simply swap and say packman 4 eats 2 and 6 because that's impossible in 4 steps. However, you can allocate food 2 to packman 2 to make it doable (and in even fewer steps).

Reply:
Maybe something along these lines? (probably in a spoiler because its wall of text) I don't know why the proof is so messy and long. If you find a cleaner version please use it. 

1) Time only depends on leftmost and rightmost pellets. Therefore, every packman will eat a contiguous, disjoint subsegment (same as before)
2) From this point on, "solution" refers to a valid assignment of packmen -> pellets that takes less than $d$ time. If there exists a solution where packmen are assigned pellets in overlapping subsegments, there exists a solution where packmen are assigned pellets in non-overlapping subsegments (assign all overlapping pellets to either packman)
3) If there exists any solution, there exists a solution where the packmen are assigned subsegments in sorted order (i.e, the $i$-th packman in sorted order is assigned the $i$-th subsegment in sorted order)

Proof 1: Two consecutive packmen are at positions $x$ and $y$, where subsegment of $x$ is to the right of subsegment $y$. The union of these subsegments is from $l$ to $r$ (and all pellets between). Assume $l \leq (x+y)/2 \leq r$. Otherwise, the claim is true by reassigning all pellets in the range to the closer packman. Since subsegment of $x$ is to the right of that of $y$, Packman at $x$ eats pellet $r$ and packman at $y$ eats pellet $l$. Thus, $max(r-x, y-l) \leq d$. We find a different pellet assignment that requires no greater than the same time $d$. Assign packman $x$ to all pellets in the range $[l, (x+y)/2]$, and packman $y$ to all pellets in the range $[(x+y)/2, r]$. Packman $x$ can travel $(x+y)/2 - x + (x+y)/2 - l = y-l$ time, and packman $y$ can travel $y-(x+y)/2+r-(x+y)/2 = r-x$ time. Thus, a solution with total time $max(y-l, r-x)$, no more than the total time $d$ of the solution, is achieved.

Proof 2: More of a thought experiment?? Number the packmen $1 \dots N$ from left to right. Simulate the packmen's movement as described in problem. However, when two packmen cross, instead switch both of their directions and not allow them to cross (similar to ant on stick problem). Notice how the overall movement and all covered regions (and thus eaten pellets) stay exactly the same. Also, this requires exactly the same amount of time. Now, packman $i$ is always to the left of packman $i+1$, so it's interval must be also to the left of packman $i+1$. 

4) Since leftmost packman must cover a leftmost subsegment, simply assign it the largest such subsegment.

Reply:
> Food at positions 0, 2, 6.

I thought all the food and packmen had to be at distinct positions (not that it really matters)

Reply:
oh, ok. well idea is still there